base/intset.jl

```diff
@@ -1,339 +1,231 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-type IntSet <: AbstractSet{Int}
-    bits::Array{UInt32,1}
-    limit::Int
-    fill1s::Bool
-
-    IntSet() = new(zeros(UInt32,256>>>5), 256, false)
+immutable IntSet <: AbstractSet{Int}
+    bits::BitVector
+    IntSet() = new(fill!(BitVector(256), false))
 end
-IntSet(itr) = (s=IntSet(); for a in itr; push!(s,a); end; s)
+IntSet(itr) = union!(IntSet(), itr)
 
 eltype(::Type{IntSet}) = Int64
 similar(s::IntSet) = IntSet()
-
-function show(io::IO, s::IntSet)
-    print(io, "IntSet([")
-    first = true
-    for n in s
-        if n > s.limit
-            break
-        end
-        if !first
-            print(io, ", ")
-        end
-        print(io, n)
-        first = false
-    end
-    if s.fill1s
-        print(io, ", ..., ", typemax(Int)-1)
-    end
-    print(io, "])")
+copy(s1::IntSet) = copy!(IntSet(), s1)
+function copy!(to::IntSet, from::IntSet)
+    resize!(to.bits, length(from.bits))
+    copy!(to.bits, from.bits)
+    to
 end
+eltype(s::IntSet) = Int
+sizehint!(s::IntSet, n::Integer) = (_resize0!(s.bits, n); s)
 
-copy(s::IntSet) = union!(IntSet(), s)
-
-function sizehint!(s::IntSet, top::Integer)
-    if top >= s.limit
-        lim = ((top+31) & -32)>>>5
-        olsz = length(s.bits)
-        if olsz < lim
-            resize!(s.bits, lim)
-            fill = s.fill1s ? UInt32(-1) : UInt32(0)
-            for i=(olsz+1):lim; s.bits[i] = fill; end
-        end
-        s.limit = top
+# An internal function for setting the inclusion bit for a given integer n >= 0
+@inline function _setint!(s::IntSet, idx::Integer, b::Bool)
+    if idx > length(s.bits)
+        !b && return s # setting a bit to zero outside the set's bits is a no-op
+        newlen = idx + idx>>1 # This operation may overflow; we want saturation
+        _resize0!(s.bits, ifelse(newlen<0, typemax(Int), newlen))
     end
+    unsafe_setindex!(s.bits, b, idx) # Use @inbounds once available
     s
 end
 
-function push!(s::IntSet, n::Integer)
-    if n >= s.limit
-        if s.fill1s
-            return s
+# An internal function to resize a bitarray and ensure the newly allocated
+# elements are zeroed (will become unnecessary if this behavior changes)
+@inline function _resize0!(b::BitVector, newlen::Integer)
+    len = length(b)
+    resize!(b, newlen)
+    len < newlen && unsafe_setindex!(b, false, len+1:newlen) # resize! gives dirty memory
+    b
+end
+
+# An internal function that takes a pure function `f` and maps across two BitArrays
+# allowing the lengths to be different and altering b1 with the result
+function _matched_map!{F}(f::F, b1::BitArray, b2::BitArray)
```

Comment:
Yeah I thought the type parameters in cases such as this used to make a difference but at some point something changed so that they didn't matter, but I don't have anything concrete to point to.